org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/references/participants/XMLReferencesDiagnosticParticipant.java

```diff
@@ -0,0 +1,99 @@
+/*******************************************************************************
+* Copyright (c) 2023 Red Hat Inc. and others.
+* All rights reserved. This program and the accompanying materials
+* which accompanies this distribution, and is available at
+* http://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lemminx.extensions.references.participants;
+
+import java.text.MessageFormat;
+import java.util.Collection;
+import java.util.List;
+
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.extensions.contentmodel.settings.XMLValidationSettings;
+import org.eclipse.lemminx.extensions.references.XMLReferencesPlugin;
+import org.eclipse.lemminx.extensions.references.search.ReferenceLink;
+import org.eclipse.lemminx.extensions.references.search.SearchEngine;
+import org.eclipse.lemminx.extensions.references.search.SearchNode;
+import org.eclipse.lemminx.extensions.references.search.SearchNode.ValidationStatus;
+import org.eclipse.lemminx.services.extensions.diagnostics.IDiagnosticsParticipant;
+import org.eclipse.lemminx.utils.XMLPositionUtility;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.DiagnosticSeverity;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.jsonrpc.CancelChecker;
+
+/**
+ * XML references diagnostics which reports undefined references.
+ *
+ */
+public class XMLReferencesDiagnosticParticipant implements IDiagnosticsParticipant {
+
+	private final static String UNDEFINED_REFERENCE_MESSSAGE = "Undefined reference ''{0}'': nothing that matches the expression ''{1}'' defines ''{2}''.";
+
+	private final static String INVALID_PREFIX_MESSSAGE = "Invalid prefix ''{0}'': that matches the expression ''{1}'' requires the ''{2}'' prefix.";
+
+	private final XMLReferencesPlugin plugin;
+
+	public XMLReferencesDiagnosticParticipant(XMLReferencesPlugin plugin) {
+		this.plugin = plugin;
+	}
+
+	@Override
+	public void doDiagnostics(DOMDocument document, List<Diagnostic> diagnostics,
+			XMLValidationSettings validationSettings, CancelChecker cancelChecker) {
+		Collection<ReferenceLink> links = SearchEngine.getInstance().searchLinks(document,
+				plugin.getReferencesSettings(),
+				cancelChecker);
+		if (links.isEmpty()) {
+			return;
+		}
+
+		for (ReferenceLink link : links) {
+			for (SearchNode from : link.getFroms()) {
+				// Validate the syntax of from node.
+				if (!from.isValid()) {
+					if (from.getValidationStatus() == ValidationStatus.INVALID_PREFIX) {
+						// Invalid prefix error
+						Range range = XMLPositionUtility.createRange(from);
+						String message = MessageFormat.format(INVALID_PREFIX_MESSSAGE, from.getValue(null),

```

Comment:
In the message you add the `to` expression. I think it might make more sense if you made it the `from` expression. What do you think?

Reply:
Sorry, I have not understood correctly. Could you give me the message that you have in your mind.

I think from is not very interesting since the from is the node wher ethere is an error, or perhaps I miss something.

Reply:
Okay. I agree that the expression for the `from` node is less interesting, we just need to modify the message to make it clear that the expression given is refering to the declaration node. Maybe we could do something like:

"Invalid reference ''{0}'': references *to declarations* that match the expression ''{1}'' require the ''{2}'' prefix."

Reply:
fixed